File: 30 day/C++/week2-ContiguousArray.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <map>
#include <unordered_map>

using namespace std;

    int findMaxLength(vector<int>& nums) {
        unordered_map<int, int> m;
        m.insert(pair<int, int>(0, -1));
        int count = 0, maxLen = 0;
        for (int i = 0 ; i < nums.size() ; i++) {
            count += (nums[i] == 1) ? 1 : -1;
            auto it = m.find(count);
            if (it != m.end()) {
                maxLen = (maxLen >= (i - it->second)) ? maxLen : (i - it->second);
            }
            else {
                m.insert(pair<int, int>(count, i));
            }
        }   
        return maxLen;
    }

int main() {
    vector<int> nums = {0,0,1,0,0,0,1,1};
    cout << findMaxLength(nums);
}   

File: 30 day/Java/week2_ContiguousArray.java
import java.util.*;

public class week2_ContiguousArray {
    public int findMaxLength(int[] nums) {
        Map<Integer, Integer> m = new HashMap<>();
        m.put(0, -1);
        int count = 0, maxLen = 0;
        for (int i = 0 ; i < nums.length ; i++) {
            count += (nums[i] == 1) ? 1 : -1;
            if (m.containsKey(count)) {
                maxLen = Math.max(maxLen, i - m.get(count));
            }
            else {
                m.put(count, i);
            }
        }
        return maxLen;
    }    
}